### PR TITLE
TST: add property-based test for pyssg

### PR DESCRIPTION
**Pull request by shanedsnyder**
In GitLab by @tylerjereddy on Sep 24, 2020, 19:14

@mdorier @carns @gshipman @nawtrey 

* this drafts a first property-based `hypothesis` test
for `py-ssg`, focused on testing the string used for the
group name when creating groups with ssg

* the good news is that I've found no evidence that strange
strings cause issues

* the bad news is that py-ssg/py-margo engine/group creation
contexts appear to have a global state that prevents launching
many of them from the same Python process? This is a problem for
this kind of `hypothesis` case testing. While we can do parametrized
tests with `pytest --fork` to separate the processes, this does not
work with `hypothesis` case generation, nor does the use of the
`pytest-subtests` appoach; open to suggestions here.. currently
when more than one test case runs in the same process we get the error
below (only the first case passes):

```
engine = <pymargo.core.Engine object at 0x2b4edb261390>, group_name = '', config = <pyssg.groups.Config object at 0x2b4edb2611d0>, addresses = ['ofi+tcp;ofi_rxm://192.168.100.178:46366'], callback = <callback.MyMembershipCallback object at 0x2b4edb261310>

    def __create_simple(engine, group_name, config, addresses, callback):
>       ssgid = _pyssg.group_create(engine.get_internal_mid(), group_name, addresses, vars(config), callback)
E       RuntimeError: ssg_group_create_mpi failed
```

* added brief documentation on running the tests

### CODE REVIEW

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Oct 6, 2020, 12:04

@mdorier @shanedsnyder @gshipman @nawtrey @carns 

Any ideas here? Interactions between Python GIL/C interface/group creation infrastructure or my Python call sequence is somehow out of order so the "teardown" is not happening properly after finished with a group?

**Comment by shanedsnyder** [comment]
In GitLab by @mdorier on Oct 6, 2020, 12:07

I think we would need more information from SSG on why the group creation failed. Shane is working on adding better error codes in SSG right now.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Oct 6, 2020, 12:29

Yeah, I've got a mostly complete cut at an SSG version that provides more useful error codes and debug logging that we should probably be able to use here to better diagnose what's going on. The biggest hurdle is that the API changed slightly (so that most function calls return an integer error code), so I'd need to probably create another branch of py-ssg to reflect those changes.

Let me try this ahead of our call tomorrow and hopefully I have a couple of new branches we could try out to figure out what exactly is going wrong.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Oct 8, 2020, 13:19

I was able to push new branches of ssg/py-mochi-ssg that should print out an error code indicating why the group failed. I haven't added error strings yet, but the error codes are enumerated in ssg.h. You can try installing these new packages like this:

```
spack install py-mochi-ssg@dev-new-ssg-api ^mochi-ssg@dev-error-codes
```

(Note that you'll have to re-pull your sds-repo)

I did get deadlock when trying out your example code, though. I get it with the new branches above, but also with the most recent tagged versions in spack. Is that something you've ran into? I'm trying to walk through the code now to see why it's hanging, but let me know if you've made any other modifications to the code you've shared. Nothing looks obviously wrong from looking at the call sequences, but I'm more familiar with the C API and don't fully understand the Python wrappers yet. I'll let you know if I find anything.

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Oct 8, 2020, 14:10

Ok, I will try to test directly from the feature branch you pushed to this repo when I get some time, along with the special `ssg` branch.

The pathological behavior may vary on different platforms/Python test dependency versions perhaps.

Just to clarify, since Python locks down the creation of new threads by default, are you certain that your C level code will interact gracefully with this global interpreter lock (GIL)? Most C-interface code will need shims to acquire and release the lock (GIL) that is held by the CPython interpreter. This is a pretty common issue, and while normally considered obvious, I do just want to double check?

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Oct 8, 2020, 14:54

If we're not sure, I can take a look at the pybind11 interface probably. In all likelihood, it may need something like this:
https://pybind11.readthedocs.io/en/stable/advanced/misc.html#global-interpreter-lock-gil

The helpful information for me, before diving in, would just be: are you likely creating new threads at the C/C++ level when I'm creating/using groups like this?

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Oct 8, 2020, 16:33

I actually wasn't aware of the Python GIL, so thanks for the suggestion. I did notice some GIL-related calls in both py-margo and py-ssg, but still don't fully grok when/how they should be used. (Sorry, as you can tell I'm still very much getting up to speed on these Python bindings myself and don't have any background in multi-threaded Python or pybind).

I've tried running a simple test like this with my new ssg/py-ssg branches, and it just hangs (as I mentioned before, it hangs with the tagged versions of ssg/py-ssg, too):

```
import sys
from pymargo.core import Engine
import pymargo.core
import pyssg
from pyssg.groups import Config, Member, Group, Callback

with Engine('ofi+sockets', pymargo.core.server) as engine:
    pyssg.init()

    pyssg.finalize()
    engine.finalize()

```

So, I just want to instantiate a pymargo engine, initialize SSG, finalize SSG, then finalize the engine. That should be a pretty simple sequence of events, but just hangs. It works fine if I get rid of the pyssg calls. Notice this is a slight variation on the code you shared that explicitly initializes SSG, skips the group create step, and reverses the shutdown order of SSG and the engine. FWIW, I get hangs for your code fragment, too, but I just wanted to simplify things for now.

In this simplified code fragment, no extra threads should be spawned, so I don't think that's the issue. Maybe there's just some other bug in our py-ssg bindings that's the issue now? It might be helpful if you can confirm whether you get the same deadlock behavior as I do with the code from above? I was under the impression you had some SSG code basically working (just not when doing multi-threaded testing), so I'm confused about whether I don't have things setup properly or if there's a bug in the bindings.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Oct 8, 2020, 16:35

To answer your other question: we do spawn threads as part of the group create process, so sounds like we'll want to make sure the bindings are properly doing GIL-related things there, but that's secondary to me getting even the simplest example to actually terminate properly.

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Oct 8, 2020, 16:41

Ah you're right, Matthieu does have some GIL handling in that interface--I figured someone would have thought of this early on. Let me check that code block now.

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Oct 8, 2020, 16:54

Yes, I reproduced the hang, it happens after checkpoint 2 below on a `skylake-gold` architecture Linux node with `master` branches all around:

```python
import sys
from pymargo.core import Engine
import pymargo.core
import pyssg
from pyssg.groups import Config, Member, Group, Callback

with Engine('ofi+sockets', pymargo.core.server) as engine:
    print("checkpoint 1")
    pyssg.init()
    print("checkpoint 2")

    pyssg.finalize()
    print("checkpoint 3")
    engine.finalize()
    print("checkpoint 4")
```

I did notice quite a few scenarios that caused hangs when I was working on these tests, but I figured it was just because the API had a very specific teardown

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Oct 8, 2020, 17:00

Great, thanks for confirming that, Tyler! That helps ease my confusion a bit. Let me try debugging more, there's got to be some issue either in the py-ssg Python module or bindings.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Oct 21, 2020, 12:46

Sorry it took me a while to circle back to this, but I found the cause of the deadlock for this particular sequence of operations. Both SSG and Margo will initialize Argobots (the library we use for threading) if they notice that it has not yet been initialized. If either of them are responsible for initializing it, they keep track of this and also shutdown Argobots as part of their shutdown procedure.

I didn't realize for pyssg in particular that SSG was being auto-initialized/finalized as part of importing the module (i.e., from pyssg's `__init.py__`). Then, when we explicitly finalize SSG in the most recent code snippet you shared (right after checkpoint 2, where we are deadlocking), we are trying to finalize Argobots even though Margo is still using it, causing the hang. The workaround for now is to not do any of the pyssg init/finalize from your code, and just let it happen as part of the module import. That works fine for me.

We could consider either getting rid of the autoinit in SSG (so the users can explicitly control the order of operations) or we could maybe even not publicly export those calls so users don't potentially get themselves into trouble. @mdorier , would any of those options make sense? I don't necessarily understand the optional autoinit feature for SSG, and don't want to mess up something you are trying to do.

**Comment by shanedsnyder** [comment]
In GitLab by @mdorier on Oct 21, 2020, 12:56

I think it would make sense to remove the auto-init feature. I didn't know ssg was initializing/finalizing Argobots, so I thought the auto-init would simply make two fewer calls for the user to think about. Sounds like I was wrong.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Oct 21, 2020, 12:59

@tylerjereddy Your initial issue was related to creating multiple groups at a time as part of your testing. I think we wanted to try to new versions of SSG/pyssg that should print error codes indicating what caused the group create to fail. Any chance you could try that again to see if we can get more details? I've tried running your code snippet with pytest, but I don't think I'm doing it right, as it fails immediately.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Oct 21, 2020, 13:07

@mdorier I think you're right that automatically initializing/finalizing SSG is nice as it saves a couple of explicit calls. I'd be fine with just not making auto-init optional (looks like you can disable it via an environment variable now) and with removing Python definitions of init/finalize functions so users can't call them and mess things up. Any objections?

**Comment by shanedsnyder** [comment]
In GitLab by @mdorier on Oct 21, 2020, 13:09

To be honest it's been too long since I touched py-ssg, so I let you make the decisions :-)

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Oct 26, 2020, 12:33

@shanedsnyder 

I'm trying to build `mochi-ssg` with your feature branch via `spack install mochi-ssg@dev-error-codes %gcc@8.3.0 ^mochi-margo@master %gcc@8.3.0` but I'm seeing:

```
==> Installing mochi-ssg
==> No binary for mochi-ssg found: installing from source
==> mochi-ssg: Executing phase: 'autoreconf'
==> mochi-ssg: Executing phase: 'configure'
==> mochi-ssg: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

3 errors found in build log:
     136      CC       src/ssg.lo
     137      CC       src/ssg-rpc.lo
     138      CC       src/lookup3.lo
     139      CC       src/swim-fd/swim-fd.lo
     140      CC       src/swim-fd/swim-fd-ping.lo
     141    src/ssg.c: In function 'ssg_group_create_pmix':
  >> 142    src/ssg.c:476:20: error: 'g_id' redeclared as different kind of symbol
     143         ssg_group_id_t g_id = SSG_GROUP_ID_INVALID;
     144                        ^~~~
     145    src/ssg.c:458:21: note: previous definition of 'g_id' was here
     146         ssg_group_id_t *g_id)
     147         ~~~~~~~~~~~~~~~~^~~~
     148    src/ssg.c:642:47: warning: passing argument 8 of 'ssg_group_create_internal' makes pointer from integer without a cast [-Wint-conversion]

     ...

     164    src/ssg.c:2649:51: warning: unused parameter 'results' [-Wunused-parameter]
     165         pmix_info_t info[], size_t ninfo, pmix_info_t results[], size_t nresults,
     166                                           ~~~~~~~~~~~~^~~~~~~~~
     167    src/ssg.c:2649:69: warning: unused parameter 'nresults' [-Wunused-parameter]
     168         pmix_info_t info[], size_t ninfo, pmix_info_t results[], size_t nresults,
     169                                                                  ~~~~~~~^~~~~~~~
  >> 170    make[1]: *** [src/ssg.lo] Error 1
     171    make[1]: *** Waiting for unfinished jobs....
     172    src/ssg-rpc.c: In function 'ssg_group_join_send':
     173    src/ssg-rpc.c:265:12: warning: 'ret' may be used uninitialized in this function [-Wmaybe-uninitialized]
     174         return ret;
     175                ^~~
     176    make[1]: Leaving directory `/ram/tmp/treddy/spack-stage/spack-stage-mochi-ssg-dev-error-codes-2y2ctjncpawspsamsuyxrd33kbscscnq/spack-src'
  >> 177    make: *** [all] Error 2

See build log for details:
  /tmp/treddy/spack-stage/spack-stage-mochi-ssg-dev-error-codes-2y2ctjncpawspsamsuyxrd33kbscscnq/spack-build-out.txt

```

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Oct 27, 2020, 10:36

Oops! Can you try again? I missed a compilation error that only pops up when SSG is built with PMIx support, but should be good now.

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Oct 27, 2020, 18:50

I get a segfault when I cherry-pick your debug branch of `py-ssg` into this feature branch and use the debug version of `ssg` proper (your special branch above) to run the tests added here:

<details>

```
=================================================================================================================================================== test session starts ====================================================================================================================================================platform linux -- Python 3.8.6, pytest-6.1.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/treddy/mochi_projects/py-ssg
plugins: cov-2.10.1, hypothesis-5.38.1
collected 1 item                                                                                                                                                                                                                                                                                                           

test/test_create.py Fatal Python error: Segmentation fault

Current thread 0x00002ae5009b6380 (most recent call first):
  File "/home/treddy/.local/lib/python3.8/site-packages/pyssg/groups.py", line 152 in __create_simple
  File "/home/treddy/.local/lib/python3.8/site-packages/pyssg/groups.py", line 165 in create
  File "/home/treddy/mochi_projects/py-ssg/test/test_create.py", line 32 in test_tcp_ssg_engine_names
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/core.py", line 548 in test
  File "/home/treddy/mochi_projects/py-ssg/test/test_create.py", line 16 in test_tcp_ssg_engine_names
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/core.py", line 625 in run
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/executors.py", line 52 in default_new_style_executor
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/core.py", line 629 in execute_once
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/core.py", line 674 in _execute_once_for_engine
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/internal/conjecture/engine.py", line 194 in __stoppable_test_function
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/internal/conjecture/engine.py", line 218 in test_function
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/internal/conjecture/engine.py", line 943 in shrink_interesting_examples
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/internal/conjecture/engine.py", line 906 in _run
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/internal/conjecture/engine.py", line 497 in run
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/core.py", line 750 in run_engine
  File "/home/treddy/.conda/envs/py38/lib/python3.8/site-packages/hypothesis/core.py", line 1131 in wrapped_test
  File "/home/treddy/mochi_projects/py-ssg/test/test_create.py", line 16 in test_tcp_ssg_engine_names
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/python.py", line 184 in pytest_pyfunc_call
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/python.py", line 1627 in runtest
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/runner.py", line 163 in pytest_runtest_call
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/runner.py", line 256 in <lambda>
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/runner.py", line 310 in from_call
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/runner.py", line 255 in call_runtest_hook
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/runner.py", line 216 in call_and_report
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/runner.py", line 127 in runtestprotocol
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/runner.py", line 110 in pytest_runtest_protocol
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/main.py", line 338 in pytest_runtestloop
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/main.py", line 313 in _main
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/main.py", line 257 in wrap_session
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/main.py", line 306 in pytest_cmdline_main
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/manager.py", line 84 in <lambda>
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/treddy/.local/lib/python3.8/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/config/__init__.py", line 164 in main
  File "/home/treddy/.local/lib/python3.8/site-packages/_pytest/config/__init__.py", line 187 in console_main
  File "/home/treddy/.local/bin/pytest", line 8 in <module>
Segmentation fault (core dumped)
```

</details>

When I switched back to the "vanilla" branches `pytest test/` complained about something else `ImportError: /home/treddy/github_projects/spack/opt/spack/linux-centos7-skylake_avx512/gcc-8.3.0/mochi-margo-master-d6gtntb37ypszfx6myrdohvi32hpkwc7/lib/libmargo.so.0: undefined symbol: hg_proc_memcpy`

I'm starting to get confused by all the fixes needed--for example, this feature branch needs a rebase for pybind11 shims and then we also need to pin pmix to master because pkgconfig issues, so it is easy to mess up an environment.

I seem to have `spack load`ed the appropriate mercury deps to provide that symbol.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Oct 29, 2020, 12:55

I've at least reproduced the segfault you mention after running `pytest test` on your fork with my cherry-picked commit. I'm investigating now to see if I can see what specifically is triggering it for these tests.

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 15:33

added 3 commits

<ul><li>c59391d8...c9d0af4b - 2 commits from branch <code>master</code></li><li>8e4a574c - TST: add property-based test for pyssg</li></ul>

[Compare with previous version](/sds/py-ssg/-/merge_requests/5/diffs?diff_id=11154&start_sha=c59391d83144ce978b07fd80d63826683a0ffba4)

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Nov 16, 2020, 16:27

FYI, I ended up pushing a couple of bug fixes to my branch of SSG (the C library, not Python bindings) that were uncovered by these tests. One of the bugs caused segfaults on an error path that we hadn't previously hit before, and the other bug caused deadlock whenever the only member of a single-member group tried to leave it.

I can get your pytest stuff running now, with one more small modification to your test program: remove the call to `pyssg.finalize()` in your `test_tcp_ssg_engine_names()` function. Otherwise, SSG is shutdown which causes any subsequent group create calls to fail. After that, the pytest commands are succeeding.

So, to summarize, you still need to check out my specific branch of SSG for the time being, and you still need to cherry-pick my modifications to py-ssg bindings, but that should be it. I will try to get these things all merged into master branches soon so we don't have to worry about keeping track of different branches.

Sorry this took so long, but hopefully this gets you further along now. Thanks again for reporting these issues as they weren't actually problems with the Python bindings, but affected our C library which is more widely used.

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 16:33

added 1 commit

<ul><li>961f8756 - TST: remove a finalize</li></ul>

[Compare with previous version](/sds/py-ssg/-/merge_requests/5/diffs?diff_id=11157&start_sha=8e4a574c6db757bde4f0ca499ea7dea205c8dcd3)

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 16:36

> So, to summarize, you still need to check out my specific branch of SSG for the time being

I was working on this just now--with your suggested adjustment to the unit test it passes just fine with `master` branch of mochi-ssg and `develop` of mochi-margo, in my hands..

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Nov 16, 2020, 16:39

I didn't try that, but that's great to know. I guess that makes sense -- we removed the problematic line of code, which avoided tripping the buggy error path. At any rate, it'll be cleared up in master branch soon enough anyways. Let me know if you hit any other issues.

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 16:41

> it'll be cleared up in master branch soon enough anyways

You mean I'll soon get a more meaningful error message with the extra finalize/initialize cycle or it will "just work" even with the extra finalize?

@nawtrey I've updated this feature branch, and now that you're getting familiar with `hypothesis`, perhaps you'll be able to confirm that these tests are now passing in your hands too (hard part is to install everything "just right" on i.e., darwin supercomputer).

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 16:44

Commented on [test/test_create.py line 42](https://github.com/mochi-hpc/py-mochi-ssg/compare/c9d0af4b3a30dfbf6705091495d01d96831ea4d8..961f8756ff6d46d5784d44766c561e2d45b2c098#diff-7c34cb71953aeef95737e4585468220cR42)



@shanedsnyder The need to call `engine.finalize()` seems a bit strange from a Pythonic standpoint--shouldn't the `engine` context manager handle the resource finalization?

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 16:45

Commented on [test/test_create.py line 42](https://github.com/mochi-hpc/py-mochi-ssg/compare/c9d0af4b3a30dfbf6705091495d01d96831ea4d8..961f8756ff6d46d5784d44766c561e2d45b2c098#diff-7c34cb71953aeef95737e4585468220cR42)



In fact, the test passes without that too!

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 16:47

Commented on [test/test_create.py line 42](https://github.com/mochi-hpc/py-mochi-ssg/compare/c9d0af4b3a30dfbf6705091495d01d96831ea4d8..961f8756ff6d46d5784d44766c561e2d45b2c098#diff-7c34cb71953aeef95737e4585468220cR42)



changed this line in [version 4 of the diff](/sds/py-ssg/-/merge_requests/5/diffs?diff_id=11158&start_sha=961f8756ff6d46d5784d44766c561e2d45b2c098#d73f5d7a3ed6115337c53a72622f15f50af03e4f_42_42)

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 16:47

added 1 commit

<ul><li>9105d9e5 - MAINT: MR 5 revisions</li></ul>

[Compare with previous version](/sds/py-ssg/-/merge_requests/5/diffs?diff_id=11158&start_sha=961f8756ff6d46d5784d44766c561e2d45b2c098)

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 16:52

added 1 commit

<ul><li>d6b7dec8 - DOC: cleanup MR 5 comments</li></ul>

[Compare with previous version](/sds/py-ssg/-/merge_requests/5/diffs?diff_id=11159&start_sha=9105d9e5cf689e7920877a612df98bfbf476e715)

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 16:54

resolved all threads

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 16:54

unmarked as a **Work In Progress**

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 16, 2020, 17:00

I'm reasonably happy with the state of this MR now, but I'm not an SSG expert of course.

One thing we could perhaps do is parametrize over more engines than just `'ofi+tcp'`; I'm open to suggestions  for that and other properties we may want to test here.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Nov 17, 2020, 10:44

It will still fail with the call to SSG finalize (at least at that point in the code -- should be fine to call that after you're done creating SSG groups), but will at least print error codes we can map to specific SSG errors.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Nov 17, 2020, 10:46

Commented on [test/test_create.py line 42](https://github.com/mochi-hpc/py-mochi-ssg/compare/c9d0af4b3a30dfbf6705091495d01d96831ea4d8..961f8756ff6d46d5784d44766c561e2d45b2c098#diff-7c34cb71953aeef95737e4585468220cR42)



You're right -- the engine should automatically deconstruct itself when it falls out of scope, so you can avoid that call. 

Similarly, the SSG group will be automatically deconstructed (left), so should be fine to forgo the `g.leave()` call you have.

**Comment by shanedsnyder** [comment]
In GitLab by @tylerjereddy on Nov 17, 2020, 11:03

Commented on [test/test_create.py line 42](https://github.com/mochi-hpc/py-mochi-ssg/compare/c9d0af4b3a30dfbf6705091495d01d96831ea4d8..961f8756ff6d46d5784d44766c561e2d45b2c098#diff-7c34cb71953aeef95737e4585468220cR42)



Removing the `g.leave()` causes the test suite to segfault in my hands.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Nov 17, 2020, 11:07

I think the idea of testing more transports is great. That's not SSG-specific functionality, but seems like something worth iterating over. There are at least a few transports we could test on most systems, but probably more than that directly on the HPC systems.

Off the top of my head, there are a few other things we could try to parameterize:
1. Groups larger than a single process.
2. More coverage of SSG API. Specifically, testing of the `observe()` function, which would require running multiple process tests (mpi4py?).
3. Try non-default values for the input `group_config` for the group create call to see if anything breaks.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Nov 17, 2020, 11:09

Commented on [test/test_create.py line 42](https://github.com/mochi-hpc/py-mochi-ssg/compare/c9d0af4b3a30dfbf6705091495d01d96831ea4d8..961f8756ff6d46d5784d44766c561e2d45b2c098#diff-7c34cb71953aeef95737e4585468220cR42)



Of course it does, that would be too easy... I'll have a look

**Comment by shanedsnyder** [comment]
In GitLab by @nawtrey on Nov 23, 2020, 18:02

mentioned in merge request !6
